File: ciomart_flutter/lib/models/sale.dart
import 'cart_item.dart';
import 'user.dart';

class Sale {
  final int? id;
  final String saleNumber;
  final User? cashier;
  final List<CartItem> items;
  final double subtotal;
  final String discountType;
  final double discountAmount;
  final double taxAmount;
  final double total;
  final double cashPaid;
  final double cashChange;
  final DateTime? saleDate;

  Sale({
    this.id,
    required this.saleNumber,
    this.cashier,
    required this.items,
    required this.subtotal,
    this.discountType = 'NONE',
    required this.discountAmount,
    required this.taxAmount,
    required this.total,
    required this.cashPaid,
    required this.cashChange,
    this.saleDate,
  });

  factory Sale.fromJson(Map<String, dynamic> json) {
    return Sale(
      id: json['id'] is int ? json['id'] : int.tryParse(json['id']?.toString() ?? ''),
      saleNumber: json['sale_number'],
      cashier: json['cashier_id'] != null 
          ? Cashier(
              id: json['cashier_id'] is int ? json['cashier_id'] : int.tryParse(json['cashier_id'].toString()),
              username: json['username'] ?? '',
              fullName: json['cashier_name'] ?? '',
            )
          : null,
      items: [], // Populated separately if needed
      subtotal: json['subtotal'] is num ? (json['subtotal'] as num).toDouble() : double.parse(json['subtotal'].toString()),
      discountType: json['discount_type'] ?? 'NONE',
      discountAmount: json['discount_amount'] is num ? (json['discount_amount'] as num).toDouble() : double.parse(json['discount_amount'].toString()),
      taxAmount: json['tax_amount'] is num ? (json['tax_amount'] as num).toDouble() : double.parse(json['tax_amount'].toString()),
      total: json['total'] is num ? (json['total'] as num).toDouble() : double.parse(json['total'].toString()),
      cashPaid: json['cash_paid'] is num ? (json['cash_paid'] as num).toDouble() : double.parse(json['cash_paid'].toString()),
      cashChange: json['cash_change'] is num ? (json['cash_change'] as num).toDouble() : double.parse(json['cash_change'].toString()),
      saleDate: json['sale_date'] != null ? DateTime.parse(json['sale_date']) : null,
    );
  }
}

File: src/main/java/com/mycompany/model/Sale.java
package com.mycompany.model;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * Class Sale - represents a completed transaction
 * Aggregation dengan User (Cashier)
 */
public class Sale {
    private int id;
    private String saleNumber;
    private User cashier; // Aggregation
    private List<CartItem> items; // Snapshot dari cart items
    private BigDecimal subtotal;
    private String discountType;
    private BigDecimal discountAmount;
    private BigDecimal taxAmount;
    private BigDecimal total;
    private BigDecimal cashPaid;
    private BigDecimal cashChange;
    private LocalDateTime saleDate;
    
    // Constructor
    public Sale() {
        this.items = new ArrayList<>();
        this.saleDate = LocalDateTime.now();
    }
    
    public Sale(int id, BigDecimal total, User cashier, LocalDateTime saleDate) {
        this.id = id;
        this.total = total;
        this.cashier = cashier;
        this.saleDate = saleDate;
        this.items = new ArrayList<>();
    }
    
    // Getters and Setters
    public int getId() { return id; }
    public void setId(int id) { this.id = id; }
    
    public String getSaleNumber() { return saleNumber; }
    public void setSaleNumber(String saleNumber) { this.saleNumber = saleNumber; }
    
    public User getCashier() { return cashier; }
    public void setCashier(User cashier) { this.cashier = cashier; }
    
    public List<CartItem> getItems() { return items; }
    public void setItems(List<CartItem> items) { this.items = items; }
    
    public BigDecimal getSubtotal() { return subtotal; }
    public void setSubtotal(BigDecimal subtotal) { this.subtotal = subtotal; }
    
    public String getDiscountType() { return discountType; }
    public void setDiscountType(String discountType) { this.discountType = discountType; }
    
    public BigDecimal getDiscountAmount() { return discountAmount; }
    public void setDiscountAmount(BigDecimal discountAmount) { this.discountAmount = discountAmount; }
    
    public BigDecimal getTaxAmount() { return taxAmount; }
    public void setTaxAmount(BigDecimal taxAmount) { this.taxAmount = taxAmount; }
    
    public BigDecimal getTotal() { return total; }
    public void setTotal(BigDecimal total) { this.total = total; }
    
    public BigDecimal getCashPaid() { return cashPaid; }
    public void setCashPaid(BigDecimal cashPaid) { this.cashPaid = cashPaid; }
    
    public BigDecimal getCashChange() { return cashChange; }
    public void setCashChange(BigDecimal cashChange) { this.cashChange = cashChange; }
    
    public LocalDateTime getSaleDate() { return saleDate; }
    public void setSaleDate(LocalDateTime saleDate) { this.saleDate = saleDate; }
    
    public LocalDateTime getTimestamp() { return getSaleDate(); }
    
    // Business method
    public void addItem(CartItem item) {
        this.items.add(item);
    }
}
